### DISCUSSION TITLE
3D Photography using Context-aware Layered Depth Inpainting

### DISCUSSION THREAD

**Discussion by sina-masoud-ansari** [thygate/stable-diffusion-webui-depthmap-script#50] [Ideas]
Hello, just wondering if there is any interest in integrating [this](https://shihmengli.github.io/3D-Photo-Inpainting/)?

I managed to get it working and it might be a cool way to let people experience their creations in simulated 3D without any additional gear

So far it only seems to work properly with midas and makes a call out to boost. I've provided an example below

If there's interest I can create a fork of their repo with the changes needed to get it working (it had some OS specific commands and the requirements was out of date)

Integrating it is likely a matter of pointing to the right files assuming the dependencies play nice.

![statue](https://user-images.githubusercontent.com/830419/208282490-74a22a74-a484-477c-b432-5d5b830689e0.jpg)

https://user-images.githubusercontent.com/830419/208282498-ad371f1d-36ae-45fc-b662-9c6c04ebd65e.mp4

**Comment by thygate** [comment]
Sorry for the late reply, I saw it on mobile and meant to answer on desktop but got caught up in things ..
So, the 3D Inpainting paper has come up a number of times in different threads and is in the pipeline. I'm not in a hurry with this one, as the pointcloud ply still needs some consideration for viewing, and the video generation has potential, but needs some thinking about how to generalize the motion without getting into deforum turf etc .. And how to integrate it in the ui, layout wise, .. suggestions are welcome !

I wanted to look at cython or numba, maybe implement that stereo generation function in cython for a speedup because reading/writing single elements in python is painfully slow.

I also still wanted to add support for manual merging of maps using boost, and I wanted to try some things to get a threejs viewer into the webui ..

So plenty to do still, ..

Also wanted to investigate Parallax Mapping..
in this spiderman game all the skyscraper inside rooms seem modeled, but are apparently depth maps : https://www.youtube.com/watch?v=eX7x1mJrQJs
different implementation in garry's mod : https://www.youtube.com/watch?v=W9LJNxqP8dA

and then there's still these 3d reconstruction papers : 
https://shunsukesaito.github.io/PIFuHD/
https://huggingface.co/spaces/Yuliang/ICON

edit:
hot off the plate :
https://github.com/openai/point-e

**Comment by sina-masoud-ansari** [comment]
No worries :) 

Wow, yeah lots in the pipeline already. Parallax mapping would be neat but I guess that needs some specific shader / webgl integration.

The 3D reconstruction stuff is incredible, didn't know that was possible!

**Reply by thygate** [reply]
Yeah, things move at an incredible pace in this field and community. So many papers, so little time :)

Someone just made an improvement to your contributed stereo generation, making the in-painting part much better it seems, I ran some experiments with numba to accelerate the pixel loops and got insane speed improvements, so expect those changes soon. (See PR https://github.com/thygate/stable-diffusion-webui-depthmap-script/pull/51)

I need to do more reading up on parallax mapping, I suppose the shader from gmod could be adapted to work with three.

The 3D reconstruction stuff is pretty amazing too, yeah. I might make the pifuhd into a separate extension, i'm not sure yet.

It would be nice to have all things depth/3D integrated into one extension, I'm just not sure how to organize it all into a ui layout.

**Reply by sina-masoud-ansari** [reply]
Oh brilliant! Look forward to testing #51 out. 

Yeah might make sense to use a separate extension. Or perhaps a library with these various tools since there's so many cool things happening in this space.

I've successfully extracted the whole stereo gen. process into a separate program so I can run SD and stereo gen. on different GPUs since it's quite time consuming to convert a batch of video frames and I wanted to use SD/VR in parallel.

So now I just need to connect the image->depthmap->stereo workflow to the video->frames extractor and frames->video process I wrote recently to get an end-to-end mono2stereo video pipeline.

The tricky part is the boost library integration and all the options there, so I'm looking at how to make that dependency cleaner. Maybe by making boost it's own library, or just extracting the necessary file dependencies. Not sure if you have any thoughts around that?

The bottleneck seems to be depthmap generation, so I think the next step is to batch the frame processing over multiple-gpus and then make it easy to deploy and run remotely, because anything at modern resolutions that's more than a few seconds is going to take ages.

**Reply by sina-masoud-ansari** [reply]
I think pulling out the boost repo files should be ok, there's 8 of them and then I can refactor the options handling and strip out other code not being used

**Reply by thygate** [reply]
I've done some short 4k clips overnight without boost, it's doable for short videos with the numba speedup. I was using ffmpeg to extract frames, audio and recombine them.

For reference :

extract audio:
`.\ffmpeg.exe -i '.\vid.mp4' -q:a 0 -map a audio.mp3`

extract frames :
`.\ffmpeg.exe -i '.\vid.mp4' frames_in\$frame%05d.png`

recombine processed frames : 
`.\ffmpeg.exe -r ORIG_FRAMERATE -i frames_out/%05d.png -i audio.mp3 -vcodec libx264 -crf 25 -pix_fmt yuv420p -acodec copy -shortest out.mp4`

Splitting the depthmap generation into a separate pass and then running the stereo generation afterwards with compiled native code will probably still be somewhat faster. (the stereo image repo has a C++ implementation of those 2 functions) It currently runs a single thread, so lots of room for improvement there.

I cloned the needed repos, boost and midas (cause boost contains old midas) into /repositories so I could not edit them, so I had to pull everything into the script that couldn't find imports or needed editing to function.
I could have just copied it all into this repo and go from there, but the initial thought was to keep the cloned repos intact, so other users/extensions can use it there too.

If I would start over, I would probably copy everything locally and edit it ..

**Reply by sina-masoud-ansari** [reply]
Excellent thanks, yeah I was using boost so that was prob. overkill for video :) res101 without boost seems acceptable performance wise. I had some success with Moviepy for managing the calls to ffmpeg, the overall process looks like this in case it is useful at some point:

```python
from functools import partial
from moviepy.editor import VideoFileClip

def as_stereo(model, pix2pixmodel, device, boost, image):
    # this calls out to the depthmap script functions, generating depth map (with or without boost)
    # and generating the stereo image
    return convert(image, model, pix2pixmodel, device, boost=boost)

device = torch.device("cuda")
boost = False

# loads a depth model e.g. res101 and merge model if boost requested
model, pix2pixmodel = load_depth_model(device, boost)

# defines function to be used with Moviepy per frame filtering method fl_image
fl_func = partial(as_stereo, model, pix2pixmodel, device, boost)

path = "mono.mp4"

with VideoFileClip(path) as video_clip:
    stereo_clip = video_clip.fl_image(fl_func)
    stereo_clip.write_videofile("stereo.mp4")

```

**Reply by thygate** [reply]
Thanks for that example, that certainly makes things quite a bit easier to implement =) I will definitely come back to this..

And yes, I agree, res101 performs fine with large enough net size for video.

We just had another contribution improving the stereo image generation technique, check it out, another significant improvement!

I will be implementing midas 3.1 next, solving the current issue with cloning an older commit of it, and bringing a few new depth models..

**Comment by AugmentedRealityCat** [comment]
> Hello, just wondering if there is any interest in integrating [this](https://shihmengli.github.io/3D-Photo-Inpainting/)?
> 

I've had loads of fun playing with this over the last two months. Here are a few links that may be of interest to you:

Version running on windows, adapted and tweaked by @donlinglok 
https://github.com/donlinglok/3d-photo-inpainting

Discussion about using this with Automatic1111 WebUI 
https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/4268

**Reply by thygate** [reply]
I have it running, and partially implemented already, but there's a shitload of options to include, and quite a few dependencies that might make the extension less compatible ..

I'll get back to that one as soon as I implement midas 3.1, I will have to put the files locally in the script folder now as the install script can't seem to manage installing a specific commit hash .. On the other hand, that means i can now edit the files so I don't have to pull a lot of code into the script itself.

**Reply by thygate** [reply]
New v0.3.3 is ready, no more cloning repos, all files contained inside extension now.

**Reply by donlinglok1** [reply]
Thanks for your masterpiece work!

I am not a 3D modeling professional, but here is some toy code i edited. If you used the 3D-Photo-Inpainting code, you can also upgrade the MIDAS to LeRes by only few steps.

here is the test code i edited.
https://colab.research.google.com/drive/1DEySGgVdZtpdy3aP0xP8Sa4yvYTxxV_y?usp=sharing

**Reply by thygate** [reply]
Thank you for your kind words.

The 3DPhoto is on the way, if you look at the script and install.py you will see some code and a lot of the dependencies are already there..

**Reply by thygate** [reply]
The 3D photo inpainting is slooooooooow ..

**Reply by thygate** [reply]
No options yet, but next version has 3D photo inpainting, I need to do a little more testing and will release asap..

**Reply by AugmentedRealityCat** [reply]
If you need help with testing, let me know, it would be a pleasure to help you in any way I can. 
But the 3d photo inpainting process is indeed slow, particularly the "now writing PLY and doing basically everything" part (I'm writing this from memory - the exact wording is probably different).

Now that you have played with the code, can you describe the difference between the 3d-photo-inpainting process and the blender depthmap input add-on ? How much "inpainting" is actually happening, and how ? Since the Blender add-on is easier to control, I've been using that to extract 3d models instead of 3d-photo-inpainting for some time now. But your next version might change all that !

Is there any way to re-use the code you have implemented for filling the gaps in stereo image pairs as inpainting color information for the 3d-photo-inpainting process? What is the upcoming version of your extension using as a 3d output format? Have you kept the basic PLY + vertex color map?

Thanks again ! Your extension + the depthmap input hack have been a revolution in the way I use Stable Diffusion.

**Reply by thygate** [reply]
I just made the commit, have at it =)

It was crashing my debugger a lot in the beginning before i realized the bilateral filter was not optional ..

I split the mesh generation and video generation into two sections .. So it's easy and quick to generate videos once you have the mesh.
The UI is a bit messy like this, not sure how to make the layout, suggestions welcome.

Initially the video generation required the rgb and depthmap and the mesh, but was able to get rid of it by writing the required value into the ply comments. It already stored original width, height, and fov like this. So now it only needs the mesh to generate videos.

I optimized the read_ply function a bit, got me a 20% increase in speed for (re-)reading the mesh.

As far as I understand it projects in the camera perspective, but segments the mesh into 'layers', whose edges are inpainted by some margin. There are still holes, but the overall freedom of movement before they become visible is increased compared to the method previously discussed / the blender addon.

You can import this mesh into blender, and use the 'Color Attribute' Node to plug into the Shader 'color' to get the mesh rendering in color, and set the camera to origin and pointing up to align them. Might also need to set the FOV to match.

It also occurred to me that this might also be a source for filling the gaps in the stereo images, but the current method is near realtime while this method requires a dozen minutes or so for a single frame .. 

Yes, the 3D format is ASCII PLY, list of vertices with rgba color and list of faces. Converting to Wavefront OBJ should be easy, but is equally bulky due to the ASCII nature. Is there such a thing as .ply.zip, or .zply, I might consider implementing something similar as the files are quite large.

Results are quite impressive, I've generated some very pleasing slow long panning shots of that spaceship image that could go straight into a movie..

**Reply by thygate** [reply]
https://user-images.githubusercontent.com/96586282/210327195-d361cade-dba7-470a-b7e4-f828e2e94c3f.mp4

original image from https://github.com/thygate/stable-diffusion-webui-depthmap-script/discussions/34 @bugmelone

**Comment by keenanamigos** [comment]
Hey all! Seems like I was tagged by accident here - just wanted to give a heads up so that correct person could be tagged 😄

**Reply by thygate** [reply]
I deleted some spam earlier today, lots of names were in it .. probably what it was ..

edit: just checked (still have the email notification), and yes your name @keenanamigos was in it .. So you can ignore the tag =)

**Reply by Thermotronica** [reply]
This is amazing

**Comment by Thermotronica** [comment]
Can i share it?

**Reply by thygate** [reply]
Did you mean to quote someone, what is it you wish to share ?

I'm pretty sure can you share anything in this thread ;)

**Comment by graemeniedermayer** [comment]
3d photo inpaint is actually insane! Great job on the integration. Even with very chaotic pictures it does a amazing job.

https://user-images.githubusercontent.com/38518120/210902013-6669c882-c3ef-49a5-b148-c1bd099840da.mp4

**Reply by thygate** [reply]
That is a great showcase for it !

**Comment by AugmentedRealityCat** [comment]
Finally got some time to test this. Just did a first full run including the 3d model and the 4 videos - it took _3m52s_ for a 1024x512 image.

(EDIT: it took that long because it downloaded the extra models for 3d-photo-inpainting - I just ran the same thing again, and this time it only took **2m43s** ! this is much faster than the standalone version !)

I got this warning along the way but everything seems to be working !  
Should I worry about the warning ? Here is the log in case this information might be helpful.
```
WARNING:py.warnings:C:\stable-diffusion-webui-master\extensions/stable-diffusion-webui-depthmap-script/scripts\inpaint\mesh_tools.py:174: RuntimeWarning: divide by zero encountered in divide
  input_disp = 1./np.abs(input_depth)
```
Now, time to check that PLY model and see if the pipeline from SD-to-blender-to-C4d  I was using with the standalone version of 3d-photo-inpainting is still working. Normally it should. EDIT: Everything works. The inpainting works - that was the main thing I was looking for ! Combined with the latest model (the 512 edition) I got amazing results. I am rendering a 30s slow-motion flyover in C4d - just a plain camera move. I'll post it later. In the meantime, here is one of the 4 automated ones:

https://user-images.githubusercontent.com/114955776/211148837-f52619c4-fd9a-4fb2-914a-eecda766b3b1.mp4

**Reply by AugmentedRealityCat** [reply]
Here is the flyover !

https://user-images.githubusercontent.com/114955776/211193460-d62bb141-7b90-4286-bfe6-5a1e4271073e.mp4

I did not fix anything manually but I did overlay a 4k bitmap texture on the vertex color map because those colors-per-vertex provided a lower-resolution source of color than the 4k bitmap. The original render of the animation sequence is also in 4k, at 60 fps - this one is half resolution (2048x1024) and half frame rate (30 fps).

**Reply by donlinglok1** [reply]
wow!🤯

**Reply by AugmentedRealityCat** [reply]
The last frame is interesting because it shows some of the limits of that technique:

1. There is a tower on the middle right side with its top detached from the land - there is a strip of sky between it and the island.
2. There is a lot of texture stretching on nearly-horizontal surfaces. 
3. The trees and many other elements are just 'painted' on the 3d surface, and are not much 3d themselves.
4. Some of the outpainting border is visible near the two largest towers at the top left - but it's much better than the first clip.

When it's moving though you hardly perceive any of those little oddities.

**Reply by SlimeVRX** [reply]
I want to reproduce the result like your video above. Are there any more detailed instructions? I want to adjust the camera movement

**Reply by benedlore** [reply]
>overlay a 4k bitmap texture on the vertex color map

@AugmentedRealityCat 
Could you describe how you accomplished this a little bit more? I don't use C4D, but do use Max, wondering about how you were able to go from vertex colors to UV space to use a 4K texture map instead of the included vertex colors in the output model. Also, how did you get it to generate a mesh in 2 minutes at that resolution? I have never seen anything close to that.

**Reply by AugmentedRealityCat** [reply]
Since the 3d model is extracted from a picture I simply project that picture back on the model as a texture map applied in camera-space. The vertex colors were kept as a layer underneath, but they are not as detailed and there are some areas showing slight color mismatch.
If you watch closely you can see that border between the projected bitmap and the vertex color in the blue of the sky near the two towers at the top, particularly towards the end of the clip.

It should be possible to do this with 3ds max, but I must admit it's been too long since I've used it to remember how. Basically it comes down to this:

background layer: per vertex colors, no UV required
foreground layer: 4k image, UV based on camera-space
layer mask: per pixel camera visibility

**Comment by thygate** [comment]
@graemeniedermayer I added your youtube video to the README, I hope you don't mind .. ?

I also linked to this thread, absolutely awesome video by @AugmentedRealityCat, I would love to add it too with your permission.

**Reply by AugmentedRealityCat** [reply]
Go ahead, I give you my permission to add my little video to the README.

Should I also share the prompt, parameters and photo I used to synthesize the base picture ? 

Would you like me to write some kind of walk-through tutorial to get something similar to my first clip, the one with the inbuilt circular camera move ? Let me know if you are interested and I'll look into this, but probably only next week though.

Here is the pic I used for IMG2IMG - I cropped it to keep only the central rocky island (the picture is public domain on wikimedia):
https://upload.wikimedia.org/wikipedia/commons/3/39/Rocher_perc%C3%A9_.jpg

**Comment by AugmentedRealityCat** [comment]
More examples from production pipeline prototypes I've been working on:

model generation from drawing

https://user-images.githubusercontent.com/114955776/211937432-922a3376-6e8e-4af7-adf4-4b16486f0583.mp4

model split into 2 separate meshes + closing mesh holes + manual inpainting + animation test

https://user-images.githubusercontent.com/114955776/211937524-dd59d304-bbaf-447f-99fe-4d8c1a8cf64f.mp4

**Reply by donlinglok1** [reply]
🤯the hamster head can move

**Comment by Absinthze** [comment]
Hi!  Is there a way to export the mesh and texture into a 3D application?

**Reply by AugmentedRealityCat** [reply]
You definitely can. The depthmap extension has an option to output a PLY mesh with colors-per-vertex (RGB Vertex Map). This older format is not supported by all applications, but you can use Blender, which is free, to import it and export it to another more common format like FBX, OBJ or ABC.

That's how the drawing of a truck above was made: I imported the model in Cinema4d via Blender, and I simply animated the camera over there as it's more flexible than trying to define camera moves with formulas directly in the Depthmap Extension video-output module.

But there are many other ways. For the Hamster plush toy animation above, I worked differently and I did not use the PLY export function at all. Instead I worked directly in Cinema 4d and I applied some kind of displacement effect based on the depthmap as an input. Basically, this either push or pull each point on the grid according to the value of the depthmap at that point. That gave me a sofa with the hamster completely stuck with it - a bit like if both were covered by a fabric on which we would have printed the colors of the objects hidden underneath. 

The next step was to manually separate those two extrusions as two separate objects. There are many ways to achieve this, but here I simply took a polygon selection tool and I selected the area covered by the Hamster and I split it from the rest. Now I had two separate models. I removed the extra polygons that were attaching the back of the Hamster toy with the sofa, and then I filled the holes behind the hamster and on the sofa.

That hole in the sofa though did not have color information - by default it would take the colors of the hamster in front of it. So I took that texture into Stable Diffusion and I inpainted the area covering the hole on the sofa so it would be seamless with the rest. And I took the result of that operation back into Cinema4d to use as a texture map on the sofa.

The last step was to adjust the geometry of the Hamster plush toy to make the back of it rounder, and to animate it very simply just to show it was indeed detached from the other object.

In the tool I imagine we'll be using in the future, most of these steps will be automated and we will get automatic 2d-to-3d scene extraction with object separation and inpainting of all hidden surfaces + automated labeling + everything ready to be animated by prompt. I wish I were a programmer !

**Reply by AugmentedRealityCat** [reply]
If you want to see something completely different, another 2d-to-3d AI pipeline I have prototyped is based on the DMT-Mesh text2mesh add-on for Blender, along with the Dream Texture add-on. This one doesn't use Automatic1111 at all - the Dream Texture add-on has its own diffusers based version of Stable Diffusion with hacks to use custom depthmaps as inputs, and the DMT-Mesh prompt-to-3d model functions are based on completely different code, with one algorithm dedicated to transform text or pictures into Point Clouds, and a second algorithm to transform that point cloud into a mesh. 

Here is the link: https://github.com/Firework-Games-AI-Division/dmt-meshes/discussions/2#discussioncomment-4733994

(I prefer not to post the video itself over here in case people think it was made with the depthmap extension - I don't want to mislead anyone.)

I wish I could get someone to adapt this to work directly in Cinema4d as this would be much more convenient for me.

**Reply by AugmentedRealityCat** [reply]
Here is yet another pipeline test: this one shows the Image-to-PointCloud and then PointCloud-to-Mesh to synthesize a post-apocalyptic pickup truck, and Dream-Texture to create a ruined city environment for a quick animation test.
https://github.com/Firework-Games-AI-Division/dmt-meshes/discussions/7

**Reply by thygate** [reply]
Again very impressive @AugmentedRealityCat , I've been away for a while, will be catching up asap.

Is WaveFront OBJ output still desirable for the inpainted mesh ?

**Reply by AugmentedRealityCat** [reply]
> Is WaveFront OBJ output still desirable for the inpainted mesh ?

**OBJ is the obvious choice**: it is supported by most 3d applications I've came across, and it's been there for a long time.

Other formats to consider would be FBX and ABC. 
FBX allows embedding of texture maps directly in the file, something OBJ doesn't.
[ABC ](https://en.wikipedia.org/wiki/Alembic_(computer_graphics))is the most recent standard, and is better than OBJ and FBX in every way I can think of, but it's not as widely supported yet.

All these formats support per-vertex colors, which is how colors is encoded inside the PLY model when you use the 3d-photo-Inpainting code. It would be useful to have the option to convert the vertex map into a UV texture, but this is not as simple as it sounds as this would require both automatic UV unwrapping and vertex color baking.

**Comment by steampunkchris** [comment]
Yes!  OBJ or FBX would be preferred =)Sent from my iPhoneOn Jan 29, 2023, at 2:23 AM, Bob Thiry ***@***.***> wrote:﻿
Again very impressive @AugmentedRealityCat , I've been away for a while, will be catching up asap.
Is WaveFront OBJ output still desirable for the inpainted mesh ?

**Reply by thygate** [reply]
@steampunkchris @AugmentedRealityCat 
Wavefront obj (with vertex colors) is now supported, enable under settings.

**Comment by RailwayFX** [comment]
Hello!
Thanks for this amazing feature!

Could you please add an option for loading custom Z-depth images? Sometimes I edit generated z-depth and it would be great to load an edited png.

Also it would be nice if you add an ability to export into a png sequence or some other lossless format, even jpg at lowest compression would be great.

Thank you!

**Reply by AugmentedRealityCat** [reply]
> Could you please add an option for loading custom Z-depth images? Sometimes I edit generated z-depth and it would be great to load an edited png.

EDIT: I thought you were trying to use Depth Maps for image generation, while you want to use them with the 3d-photo-inpainting code. Please disregard what follows as it does not apply to your case.

----------------------

You can do that already but with a different extension, the Depth Image IO from @AnonymousCervine available here:
https://github.com/AnonymousCervine/depth-image-io-for-SDWebui

Have fun ! 
By the way, you can also use them with the new ControlNet extension as there is a model that works from depth maps as well:
https://github.com/Mikubill/sd-webui-controlnet
With the 2.0 depth model any depth map you use will be downscaled to 64x64, but with ControlNet you can use the full resolution to drive image synthesis.

As for your question regarding sequences, there is a batch tab in the IMG2IMG page and from there you can work with image sequences.

**Reply by AnonymousCervine** [reply]
I could be missing something, but I'm not quite sure my extension would help in this instance? I very-narrowly target the pipeline meant for the Stable Diffusion 2.0 depth2img model (and anything based on it/structured the same way). The "3D photo inpainting" discussed in this thread, by contrast, is a project wholly independent from Stable Diffusion that uses it's own, fundamentally-different neural net algorithm(s) and training (and it just happens to be *really cool* to use alongside SD, which I presume is why it is incorporated as a tool, here)—all of which is to say, my code doesn't even know anything about it and won't touch it, alas.

(And I believe ControlNet, while much more broadly applicable, will still run afoul of the same fundamental "this isn't Stable Diffusion or even similar to it" problem.)

**Reply by AugmentedRealityCat** [reply]
@AnonymousCervine  I just edited my comment accordingly - I was mistaken and I thought @NameRX simply wanted to use customized depth maps for image generation. My error ! Sorry, and thanks for taking the time to clear this up.

**Reply by RailwayFX** [reply]
I decided to make a mod by myself, I added:

"Load Custom DepthMap" option, it supports 16bit grayscale png format
loseless avi-png export format,
"3D inpainted mesh settings" for mesh generation
process bar for rendering frames in console
processing time in console
configs logging to txt
"Generate All button", that makes z-depth, 3d mesh and video by a single click.

![image](https://user-images.githubusercontent.com/24415698/221959739-c17e5ea6-0825-4db1-a536-3620f203b07c.png)

i'm not a professional programmer, and I have never written python code before, so it may be imperfect and buggy, sorry about that. but it works :D
[mod.zip](https://github.com/thygate/stable-diffusion-webui-depthmap-script/files/10854183/mod.zip)

**Reply by metzo007** [reply]
"Load Custom DepthMap" option, it supports 16bit grayscale png format sounds great to fix little errors in the automatic generation. Sadly the MOD dont wokr for me. If I replace the two files Auto111 will no start.

**Reply by AugmentedRealityCat** [reply]
> Sadly the MOD dont wokr for me. If I replace the two files Auto111 will no start.

Same thing for me when I tried it a couple of days ago. I tried two times, and both time it would not launch at all. 
I did not report anything because I did not even have an error message to share or anything like that.

**Reply by AugmentedRealityCat** [reply]
> i'm not a professional programmer, and I have never written python code before, so it may be imperfect and buggy, sorry about that. but it works :D [mod.zip](https://github.com/thygate/stable-diffusion-webui-depthmap-script/files/10854183/mod.zip)

I never managed to get your previous version to work but I wanted to check with you if you had kept on developing it. It would be a very useful tool.

**Reply by SlimeVRX** [reply]
I also look forward to this job

**Reply by RailwayFX** [reply]
> > i'm not a professional programmer, and I have never written python code before, so it may be imperfect and buggy, sorry about that. but it works :D [mod.zip](https://github.com/thygate/stable-diffusion-webui-depthmap-script/files/10854183/mod.zip)
> 
> I never managed to get your previous version to work but I wanted to check with you if you had kept on developing it. It would be a very useful tool.

Hi! 

What system do you have? I run it on windows 10 with Python 3.10.6, DepthMap v0.3.8. 

here is entire directory of my mod, please save original files, and try to replace with this one. Files are located in \stable-diffusion-webui\extensions\

[stable-diffusion-webui-depthmap-script.zip](https://github.com/thygate/stable-diffusion-webui-depthmap-script/files/11042418/stable-diffusion-webui-depthmap-script.zip)

**Reply by SlimeVRX** [reply]
Thank you! I will try it

**Reply by AugmentedRealityCat** [reply]
> 
> What system do you have? I run it on windows 10 with Python 3.10.6, DepthMap v0.3.8.
> 

Windows 10, python: 3.10.6 and DepthMap v0.3.8 as well. But maybe I was not a couple of weeks ago when I tried to run this. Thanks for sending the zip - I'll try that soon and I'll give you some feedback.

Have you seen the new ZoeDepth extension for Automatic1111 ? It's great really - you should try it if you haven't already, it might give you some ideas or solutions. The 3d panorama extraction is a really nice trick as well - I used to do that manually, but this is so much faster !

https://github.com/sanmeow/a1111-sd-zoe-depth

**Comment by RailwayFX** [comment]
Here are my results:

https://user-images.githubusercontent.com/24415698/221980348-a1d859a3-483a-4eb3-86c8-f46f21bd8cfe.mp4

A still picture was generated in midjorney, then I changed it a little bit using img2img in SD, then I made several upscaled pictures using SD Upscale and combined them in photoshop with Topaz Gigapixel output to fix some bad areas. After that I fixed overstretched edges at last frame by adding some details using img2img, and after that reprojected fixed frame to whole video using Ebsyth. That made a clear good looking footage at all frames. I combined all outputs in After Effects and to make it look better I added some digital compositing like moving clouds and shadows from them. Final output is in 4K.

Easy!

**Reply by AugmentedRealityCat** [reply]
Superb ! I love it :)

I'll give your code a try, it looks very interesting. And of course I'll give you feedback about how it goes !

As a side note, if anyone else has problems watching the video, right-click, `save video as`, and then open it with VLC. 
My own problems with it were caused by Windows wanting me to buy an HEVC codec for playing this directly on the webpage, while VLC does it for free. Thanks VLC !

**Reply by LanceGao97** [reply]
> I'll give your code a try, 
Hello AugmentedReal,    I see you are very active on DepthMap,    I have tried everything to batch render videos , all failed ,  all online googleColab and other servers they can only make PLY OBJs  they all fail make videos,      I  want to batch render videos by input images, 
how am I suppose to do that?  
     I remember there is a thread you mentioned   https://github.com/donlinglok/3d-photo-inpainting,
but I can't find that thread,  and I m not a scriptor , I don't know how to do it.  
Now I got 100+ PLY files ,  I want do video editiing,  
is it possible to batch render them by 1 click?     can you give me a small script of  "how to batch render videos from PLY or OBJ files"  
according to    https://github.com/donlinglok/3d-photo-inpainting,

**Reply by donlinglok1** [reply]
@LanceGao97 
try this, not guarantee still working. I didn't focus on this for a while, i think AugmentedRealityCat may have new information.

https://colab.research.google.com/drive/1DEySGgVdZtpdy3aP0xP8Sa4yvYTxxV_y?usp=sharing

**Reply by AugmentedRealityCat** [reply]
I never managed to get the code shared by @NameRX to work when I tried it and I haven't tried again since.
Everything I did with 3d-Photo-inpainting on Windows was with the help of @donlinglok - he is the programmer who knows about these things and who made it work for me before it was adapted for this extension. Now I use the extension to do those things instead.

**Reply by LanceGao97** [reply]
> @LanceGao97 try this, not guarantee still working. I didn't focus on this for a while, i think AugmentedRealityCat may have new information.
> 
> https://colab.research.google.com/drive/1DEySGgVdZtpdy3aP0xP8Sa4yvYTxxV_y?usp=sharing
Thank you， you are the creator,   but im too noob im not sure I can do this,  I m checking all the words on it.

**Reply by LanceGao97** [reply]
![图片](https://github.com/thygate/stable-diffusion-webui-depthmap-script/assets/125782753/41f2eefb-8ccc-413b-b515-8ace23d108a8)

It says  this  under  that excute line， 
Traceback (most recent call last):
  File "/content/3d-photo-inpainting/main.py", line 6, in <module>
    import vispy
ModuleNotFoundError: No module named 'vispy'

**Reply by LanceGao97** [reply]
> I never managed to get the code shared by @NameRX to work when I tried it and I haven't tried again since. Everything I did with 3d-Photo-inpainting on Windows was with the help of @donlinglok - he is the programmer who knows about these things and who made it work for me before it was adapted for this extension. Now I use the extension to do those things instead.

OK thanks for reply ,  I already got the PLY files,  I killed many hours finding ways to batch render them into videos. I m still struggleing,   I can only do basic copy paste, I don't know any python  ....  This is a nice Wolrd. Struggle always.

**Reply by donlinglok1** [reply]
> ![](https://user-images.githubusercontent.com/125782753/245385215-41f2eefb-8ccc-413b-b515-8ace23d108a8.png)
> 
> It says this under that excute line， Traceback (most recent call last): File "/content/3d-photo-inpainting/main.py", line 6, in import vispy ModuleNotFoundError: No module named 'vispy'

Seems the library is too old and need many upgrade to run, even the BoostingMonocularDepth didn't update for 2 year. 
You may need to look for other example.

**Reply by LanceGao97** [reply]
> > ![](https://user-images.githubusercontent.com/125782753/245385215-41f2eefb-8ccc-413b-b515-8ace23d108a8.png)
> > It says this under that excute line， Traceback (most recent call last): File "/content/3d-photo-inpainting/main.py", line 6, in import vispy ModuleNotFoundError: No module named 'vispy'
> 
> Seems the library is too old and need many upgrade to run, even the BoostingMonocularDepth didn't update for 2 year. You may need to look for other example.

 THank you anyway for reply,  I think it s time to give up for me,   I tried 4 on line server ,  all fail to render videos,  to be wise, is just render it with my local 4070ti,  and don't think about it anymore,  other wise, I might got lost and frastrated.    THanks again.

**Reply by charchit7** [reply]
> Here are my results:
> 
>  Mountains_AI_01_2.mp4 
> A still picture was generated in midjorney, then I changed it a little bit using img2img in SD, then I made several upscaled pictures using SD Upscale and combined them in photoshop with Topaz Gigapixel output to fix some bad areas. After that I fixed overstretched edges at last frame by adding some details using img2img, and after that reprojected fixed frame to whole video using Ebsyth. That made a clear good looking footage at all frames. I combined all outputs in After Effects and to make it look better I added some digital compositing like moving clouds and shadows from them. Final output is in 4K.
> 
> Easy!

Hey amazing work @NameRX could you please share your code to replicate it>
